Get the transform docs into the API pages.

rpgoldman opened the pull request:
I'm marking this as WIP, because the entries for the transform instances in the docs are not formatted as well as I would like.

Someone should also vet my additional exports. I found that the transforms were not being exported from distributions into pymc3 -- some of them were dropped and only exported from distributions.

If we want to have a hierarchical namespace like this, we should probably spend a little time deciding how it will be laid out.

rpgoldman committed:
Extend exports from transforms.

rpgoldman committed:
Add docs for distributions.transforms.

The doc strings for the transforms were not linked into the documentation page for distributions.

ColCarroll commented on pymc3/distributions/__init__.py:
```diff
@@ -81,11 +81,7 @@
-from .transforms import logodds
-from .transforms import log
-from .transforms import sum_to_1
+from .transforms import *
```
I like this better than the current method. Two things:

1. This is technically a breaking change, since `pm.distributions.transform` will no longer work. I bet this messes up 0 people, but we should add a deprecation warning before removing.
2. As a suggestion, I think we should surface transforms as `pm.transforms`, so that you could use `pm.transforms.ordered()` instead of `pm.distributions.ordered()` (In fact, I think right now `pm.log` might be a transform, and not a logarithm?). I _think_ we would do this by not importing anything from `.transforms` here, but in the top level init, `import .distributions.transforms as transforms`. How does that sound?

rpgoldman replied:
WRT your (1), I think we could just not export `transform`, since it's just an alias for `Transform`.  That's also theoretically a breaking change, but seems even less likely to cause troubles.  Especially since one should not use `transform` as the name of a class in python.

(2) Sounds good to me.

LMK what you think about my responses and I will make the changes in the PR, or I believe you could do them yourself, if that's what you prefer.

ColCarroll replied:
Both of those sound good to me. Feel free to ping after you mess with those imports I'll kick the tires locally (and add tests if I find anything :) ) and then merge. Ooh, I'll check the doc thing you just mentioned, too.

rpgoldman replied:
Wait -- why is `pm.distributions.transform` no longer going to work?  It's in the `__all__` list in `transforms.py`, so I think it's still ok, isn't it?
But I like the idea of deprecating it, since for a class `Transform` is preferable.  Also `transform`, not being an initialized entity, doesn't serve the same purpose as the other lower-case names.

rpgoldman commented:
When I look at the formatted docs for the transforms, the parameters are displayed like this
```
**foo**type
```
instead of
```
***foo** : type
```
as they should.  Please have a look and see if this works for you -- it could be just a local fail, in which case we can ignore it.

I really don't know what to do about what sphinx is doing with the `::autodata::`.

rpgoldman commented:
Actually, my autosummary also seems broken.

rpgoldman committed:
Improve documentation of transform aliases.

The transforms code has lower-case alternatives to the upper-cased
class names.  Sometimes these are instances of the classes (e.g.,
`log` is an instance of the `Log` transform), but sometimes they are
aliases (e.g., `lowerbound` is an alias for `LowerBound`).

I couldn't find a good way to document these in the source file, so
they are hand-documented in transforms.rst.

rpgoldman committed:
Export improvements requested by @ColCarroll.

rpgoldman committed:
Minor edits to clarify transformations.

rpgoldman committed:
Added API page for random variables.

twiecki commented on docs/source/developer_guide.rst:
```diff
@@ -378,25 +378,27 @@ https://github.com/pymc-devs/pymc3/blob/6d07591962a6c135640a3c31903eba66b34e71d8
 have a partial implementation called ``Transform``, which implements
 Jacobian correction for forward mapping only (there is no Jacobian
-correction for inverse mapping). The use case we considered are limited
+correction for inverse mapping). The uses case we considered are limited
```
use cases?

rpgoldman replied:
Thanks, Thomas!  Fixed now.

rpgoldman committed:
Fix typo @twiecki caught and some others.

Thomas caught a typo of mine, and when fixing it, I picked off some grammar errors and improved some typesetting.

twiecki commented:
Did you compile things locally to ensure everything looks right?

twiecki commented:
And needs a line in release-notes.

rpgoldman commented:
> Did you compile things locally to ensure everything looks right?

I did, but .... they don't.  See my discussion with @ColCarroll  above:

1. The `autosummary` doesn't work for me.  This is probably at least partly due to my hand-documenting the transform aliases that are exported. I do not know how to fix this. If anyone else does, that would be great.  Otherwise, I can hand-hack a `toctree` here.
2. Parameters are not properly typeset for me.  I have seen this before, but don't know what causes it.  I don't know whether this is a problem for others, or something wrong with my sphinx set up.  For me it seems to be all over the PyMC3 docs, not just here.  Here's an example from the distributions api pages:

![image](https://user-images.githubusercontent.com/3274/69556611-a35db980-0f6a-11ea-9e8c-2d88f5dbbb81.png)

rpgoldman committed:
Add to release notes.

rpgoldman commented:
> And needs a line in release-notes.

Done!

rpgoldman committed:
Fix typesetting of parameters.

Fix typesetting of parameters in the docstrings.

rpgoldman committed:
Add docstring for t_stick_breaking()

Thanks to @aloctavodia for help.

rpgoldman committed:
Slightly improved Table of Contents.

I still don't know how to get the title of *this* page *not* to appear in its table of contents.
Unfortunately, `autosummary` does not work at all.

rpgoldman committed:
Fix formatting of additional parameters.

While I was at it, fixed the punctuation of parameters in the docstrings so that Sphinx will format them correctly.

rpgoldman commented:
I've taken this as far as I can go with it.  `.. autosummary::` doesn't work because of something to do with the aliases that I had to document by hand (couldn't figure out how to make Sphinx do that properly).

I'm going to take off the "WIP" -- @ColCarroll will you please have one last look and then squash merge?

ColCarroll commented:
Built docs here -- I found the transform docs, but would love if you confirm they're what you expected!

https://pymc3.colindcarroll.com/

rpgoldman commented:
That looks as I had expected, thanks, @ColCarroll

ColCarroll commented:
thanks @rpgoldman -- careful and valuable work as always.

rpgoldman commented:
Thanks for the kind words -- couldn't have done it with the help of a bunch of the devs' kind help.
